cruise-control/src/main/scala/com/linkedin/kafka/cruisecontrol/executor/ExecutorUtils.scala

```diff
@@ -85,38 +87,126 @@ object ExecutorUtils {
+                  if (currentReplicaAssignment.contains(destinationBroker) && !currentReplicaAssignment.contains(sourceBroker)) {
+                    // Reassignment is done already.
+                    addTask = false
                   }
-                } else {
-                  if (currentReplicaAssignment.contains(sourceBroker)) {
-                    // The destination broker is already in the list, we just need to filter out the source broker if
-                    // it exists.
-                    currentReplicaAssignment.filter(_ != sourceBroker)
-                  } else {
-                    // If the source broker is no longer in the list, just do not add the task to the reassignment.
+                  // Get new replicas for replica movement.
+                  getNewReplicasForReplicaMovement(currentReplicaAssignment, task)
+                } else if (task.proposal.balancingAction() == BalancingAction.REPLICA_ADDITION) {
+                  if (currentReplicaAssignment.contains(destinationBroker) && !currentReplicaAssignment.contains(sourceBroker)) {
+                    // Reassignment is done already.
+                    addTask = false
+                  }
+                  // Get new replicas for replica addition.
+                  getNewReplicasForReplicaAddition(currentReplicaAssignment, task)
+                } else if (task.proposal.balancingAction() == BalancingAction.REPLICA_DELETION) {
+                  if (!currentReplicaAssignment.contains(destinationBroker) && !currentReplicaAssignment.contains(sourceBroker)) {
+                    // Reassignment is done already.
                     addTask = false
-                    currentReplicaAssignment
                   }
+                  // Get new replicas for replica deletion.
+                  getNewReplicasForReplicaDeletion(currentReplicaAssignment, task)
+                } else {
+                  throw new RuntimeException(s"Unexpected balancing action for proposal ${task.proposal} for [$topic, $partition]")
                 }
               }
             }
         }
         if (addTask)
-          newPartitionAssignment += (tp -> newReplicas)
+          newReplicaAssignment += (tp -> newReplicas)
       })
 
       // We do not use the ReassignPartitionsCommand here because we want to have incremental partition movement.
-      if (newPartitionAssignment.nonEmpty)
-        zkUtils.updatePartitionReassignmentData(newPartitionAssignment)
+      if (newReplicaAssignment.nonEmpty)
+        zkUtils.updatePartitionReassignmentData(newReplicaAssignment)
+    }
+  }
+  def getNewReplicasForReplicaAddition(currentReplicaAssignment: Seq[Int], task: ExecutionTask): Seq[Int] = {
+    val sourceBroker = task.sourceBrokerId()
+    val destinationBroker = task.destinationBrokerId()
+    if (!currentReplicaAssignment.contains(destinationBroker)) {
+      if (currentReplicaAssignment.contains(sourceBroker)) {
```

Comment:
For replica addition, the source should be null. So this branch should just be if source is not null throw exception. If we trust the sanity check we have already made in the constructor of BalancingProposal, we can just skip this check and assume the source is null.